successfully integrated and tested Ecotrons ECU receiver, cleaned up

silviu-toderita opened the pull request:
-Integrated Ecotrons ECU "sensor"
    -Changed "onserialdata" function in Sensor_ECU library to "handle", and it's running in every loop in the main code. This will continuously check for 27 bytes of data in the UART RX buffer and update the data when that threshold is met

-Deleted Sensor_RSSI library as this was just a demo

-Deleted a bunch of test code from main.cpp

-Not currently sure how the DataQueue library works, however it is dependent on another library that's not in the repo (or not created yet). Have commented out lines that wouldn't compile and we can come back to it later

silviu-toderita commented:
Looks like some of the automatic build checks failed. Not sure why, because I only updated Proto and not Urban. Perhaps an issue with firmware version?

silviu-toderita commented:
See comments above, except now I've fixed the compile issue. CI is cool!

lcfyi commented on lib/DataQueue/src/DataQueue.cpp:
```diff
@@ -18,12 +18,12 @@ void DataQueue::add(String id, float value) {
 
 void DataQueue::init() {
-	publishQueue = new PublishQueueAsyncRetained(publishQueueRetainedBuffer, uint16_t(sizeof(publishQueueRetainedBuffer)));
+	//publishQueue = new PublishQueueAsyncRetained(publishQueueRetainedBuffer, uint16_t(sizeof(publishQueueRetainedBuffer)));
```
> Not currently sure how the DataQueue library works, however it is dependent on another library that's not in the repo (or not created yet). Have commented out lines that wouldn't compile and we can come back to it later

Have you confirmed whether you have the submodule cloned? Unfortunately, submodules don't clone by default so if you don't initialize them, they'll be empty.

silviu-toderita replied:
Yes, that was the issue! Fixed now and I've uncommented these lines again

lcfyi commented on lib/SparkFun_GNSS/LICENSE.md:
```diff
@@ -0,0 +1,55 @@
+Adapt — remix, transform, and build upon the material
+for any purpose, even commercially.
+The licensor cannot revoke these freedoms as long as you follow the license terms.
+Under the following terms:
+
+Attribution — You must give appropriate credit, provide a link to the license, and indicate if changes were made. You may do so in any reasonable manner, but not in any way that suggests the licensor endorses you or your use.
+ShareAlike — If you remix, transform, or build upon the material, you must distribute your contributions under the same license as the original.
+No additional restrictions — You may not apply legal terms or technological measures that legally restrict others from doing anything the license permits.
+Notices:
+
+You do not have to comply with the license for elements of the material in the public domain or where your use is permitted by an applicable exception or limitation.
+No warranties are given. The license may not give you all of the permissions necessary for your intended use. For example, other rights such as publicity, privacy, or moral rights may limit how you use the material.
+
+
+Code
+--------
+
+**SparkFun code, firmware, and software is released under the MIT License(http://opensource.org/licenses/MIT).**
+
+The MIT License (MIT)
+
+Copyright (c) 2016 SparkFun Electronics
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
```
Would it be possible to include this library as a submodule as well? It would be a good candidate if you didn't have to make any changes to the original library (from [here](https://github.com/sparkfun/SparkFun_u-blox_GNSS_Arduino_Library), I assume) in order to get this to compile.

silviu-toderita replied:
Yes, I've added it as a submodule and I've also done that for the MAX31855 library. Submodules are cool, I've never used them before

lcfyi commented on urban/src/main.cpp:
```diff
@@ -69,6 +65,6 @@ void loop() {
 void onSerialData()
 {
-    ecu.onSerialData();
+    ecu.handle();
```
It's not related to this PR but I just noticed we have the ECU handlers in the urban code. I suppose it was left over from testing but that definitely shouldn't be there 😛. Good catch on removing the RSSI stuff so this'll compile, though.

silviu-toderita replied:
Fixed!

lcfyi commented on lib/Sensors/src/Sensor_ECU.cpp:
```diff
@@ -29,7 +29,7 @@ void Sensor_ECU::flush()
 }
 
-void Sensor_ECU::onSerialData()
+void Sensor_ECU::handle()
```
Slightly out of scope of this PR, but this library didn't verify the checksum when it was written; that could be an improvement down the line to verify the integrity of the data.

silviu-toderita replied:
I'll open an issue for this

lcfyi reviewed (commented):
Looks good, just one more thing outside of comments on specific lines:

- It looks like the `Particle-GPS` and `PublishQueueAsyncRK` submodules were deleted; was there a reason for the latter?

silviu-toderita commented:
Thanks for looking this over Leslie, I've fixed all the issues you found and I'll open an issue regarding the ECU data checksum. I'm going to be offline for a couple of days now, but I'm planning to get back to working on more sensors on Tuesday.

silviu-toderita commented:
Also the reason I deleted the GPS library is because it seems to only work with communication over UART, and we have switched to I2C. That's what the new SparkFun GNSS library is going to be for. If there's any functionality that you think we need that's specific to the previous library, we can always add it back in.

silviu-toderita commented:
Created a sensor "interface" and simplified Sensor_ECU library

lcfyi commented on .gitmodules:
```diff
@@ -1,6 +1,12 @@
 [submodule "lib/PublishQueueAsyncRK"]
 	path = lib/PublishQueueAsyncRK
-	url = https://github.com/rickkas7/PublishQueueAsyncRK.git
+	url = git@github.com:rickkas7/PublishQueueAsyncRK.git
```
Just a small thing, but since these are public repos, could you change these to point to the HTTPS clone URL? As it stands, the submodules will not clone from GitHub unless your SSH key is explicitly set.

It's not a huge issue for the most part, but it would save a troubleshooting step in case somebody in the future tries to clone without an SSH key registered to their account.

silviu-toderita replied:
Fixed!

lcfyi commented on lib/Sensor/src/Sensor_ECU.cpp:
```diff
@@ -0,0 +1,81 @@
+    {
+        // Data is valid
+        this->_time = Time.now();
+        fields["rpm"] = this->_interpretValue(buffer[6], buffer[7], 0.25, 0.0);
+        fields["map"] = this->_interpretValue(buffer[8], buffer[9], 0.0039, 0.0);
+        fields["tps"] = this->_interpretValue(buffer[10], buffer[11], 0.0015, 0.0);
+        fields["ect"] = this->_interpretValue(buffer[12], buffer[13], 1, -40.0);
+        fields["iat"] = this->_interpretValue(buffer[14], buffer[15], 1, -40.0);
+        fields["o2s"] = this->_interpretValue(buffer[16], buffer[17], 0.0048, 0.0);
+        fields["spark"] = this->_interpretValue(buffer[18], buffer[19], 0.5, 0.0);
+        fields["fuelpw1"] = this->_interpretValue(buffer[20], buffer[21], 0.001, 0.0);
+        fields["fuelpw2"] = this->_interpretValue(buffer[22], buffer[23], 0.001, 0.0);
+        fields["ubadc"] = this->_interpretValue(buffer[24], buffer[25], 0.00625, 0.0);
+    }
+    else
+    {
+        this->flush();
+    }
+}
+
+float Sensor_ECU::get(String field) {
+    if(fields.find(field) != fields.end()) {
+        return fields[field];
+    }
+
+    return 0;
+}
```
I'm a huge fan of this, but something worth considering is the extra overhead when using STL features in an embedded context. Have you verified the program sizes and heap usage before and after using `map`?

silviu-toderita replied:
I did not, and that's a good point. I was trying to build an interface that would have similar functions across all the sensor classes, but I'm actually a little lost on whether that would end up being useful or not or how to implement that. I will revert these changes and maybe we could discuss more at the next meeting.
